### PR TITLE
Area, Bar, and Line set their legend props through context

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

These three have to be changed in one PR because they could appear in a ComposedChart together and the Legend would be confused.

There are changes this time - the legend renders in a different order, and if a legend content is passed as a custom function then that function is called 3 times instead of once. I imagine that will get better once we remove the element cloning.

## Related Issue

https://github.com/recharts/recharts/discussions/3717

## Motivation and Context

No more element cloning!

## How Has This Been Tested?

npm test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/cartesian/Area.tsx [hunk 1]
```diff
@@ -104,6 +106,29 @@ type AreaComposedData = ChartOffset & {
   isRange: boolean;
 };
 
+function getLegendItemColor(stroke: string | undefined, fill: string): string {
+  return stroke && stroke !== 'none' ? stroke : fill;
+}
```
Comment (ckifer) [inline src/cartesian/Area.tsx:111]: nit: could this be a common function?

File: test/component/Legend.spec.tsx [hunk 2]
```diff
@@ -412,7 +412,7 @@ describe('<Legend />', () => {
     });
 
     it('should pass parameters to the function', () => {
-      expect.assertions(1);
+      // expect.assertions(1);
```
Comment (ckifer) [inline test/component/Legend.spec.tsx:415]: nit: can remove this entirely?

**Review by ckifer — APPROVED** [review]
This is huge 🎉 🎉 🎉

File: test/component/Legend.spec.tsx (hunk 2 above)
Comment (PavelVanecek) [reply test/component/Legend.spec.tsx:415]: Ah yes I will remove

File: src/cartesian/Area.tsx (hunk 1 above)
Comment (PavelVanecek) [reply src/cartesian/Area.tsx:111]: As I am untangling the generateCategoricalChart I am finding more and more unique behaviour of each graphical component so I am by default putting everything into separate function. Later once it's done there indeed might be an opportunity to extract some of these as shared code.

**Pull request merged**